### ISSUE TITLE
Improve Java2d renderer

### ISSUE THREAD

**Issue by relu91** [nifty-gui/nifty-gui#284]
I've opened an other issue to keep track of what should be changed in Java2d renderer. This need came up after the #277 pull request and our will to keep this renderer. 
In my opinion a little thing that needs refactor is font rendering, right now it's only supported stardard awt font and not bitmap fonts which is a standard de facto in the most of game engine renderers.

**Comment by relu91** [comment]
Also this issue could be a start : #231

**Comment by bgroenks96** [comment]
I would like to rewrite some of the base rendering code, like using a VolatileImage as the buffer canvas instead of BufferedImage. In my past experience, VIs often provide hardware accelerated rendering far more reliably. Conversely, they tend to perform as well and often much better than simple BI back buffers.

**Comment by ghost** [comment]
One issue that sticks out like a sore thumb:

The Java2D renderer can't use the batch rendering system, which has all sorts of nifty improvements (quad batching, multiple texture atlases, non-atlas textures, jglfont backend, font caching, image caching, etc).

It's kind of a separate issue, but we really need a new `RenderDevice` interface. The current one is so outdated that basically `BatchRenderDevice` is just an adapter to `BatchRenderBackend`. The `BatchRenderBackend` interface should be merged somehow into `RenderDevice` to create an accurate representation of the rendering system in Nifty. `BatchRenderDevice` works, but it's a convoluted mess right now.

The problem is that `BatchRenderDevice` is a concrete implementation, but contains a lot of "hidden" functionality that should be automatically implemented internally for any implementation of `RenderDevice`. This means that everyone is tied into the concrete `BatchRenderDevice` implementation, which is bad because when someone comes along and writes a new implementation of `RenderDevice`, it's going to either 1) be missing all the features I listed above, or 2) it's going to have to duplicate / rewrite them.

I see this as the root issue that is really going to hold a good Java2D renderer implementation back. It's between a rock and a hard place because you also can't tie in to the batch rendering system because that requires direct access to OpenGL: implementing the GL and / or CoreGL interfaces. The whole purpose of Java2D is to abstract away OpenGL.

This problem is why the current Java2D renderer doesn't use the jglfont library, and contains font files that are duplicated from `nifty-renderer-lwjgl` (and even those are only used by the deprecated `LwjglRenderDevice`).

Notice that every direct `RenderDevice` implementation is marked as deprecated (besides the `BatchRenderDevice` adapter). That's because the `RenderDevice` interface itself is deprecated, and needs to be marked as such.

I propose that we deal with the root issue first before proceeding, so that we can do it right the first time.

**Comment by relu91** [comment]
Sorry for my ignorance but bound nifty render engine to openGL is a limitation of freedom that it has before,isn't it ? I mean Is this  not causing issue if I'm going to write a renderer that won't use openGl at all? For example porting nifty to c# and use XNA framework. Just my random thoughts ...

**Comment by bgroenks96** [comment]
@3xp0n3nt So you're suggesting that we worry about refactoring the Nifty core API first, before we go about rewriting the Java2D implementation?

Correct me if I'm wrong, but the BatchRenderDevice is meant to write textures to TextureAtlases and ultimately draw the entire GUI in as few GL draw calls as possible correct?  While it's true that we can't make direct OpenGL calls from Java2D, it is conceivable that we can design a system using properly managed BufferedImages and (more importantly) VolatileImages in which a similar batched process takes place, at least as far as the video hardware is concerned.  It should mainly be focused on keeping making as few draw calls on the VI stored in VRAM as possible, possibly storing image data in either secondary VIs (like texture atlases) and minimizing the amount of data that ends up being shipped over the hardware bus from Java system memory to allocated video memory.

**Comment by void256** [comment]
Both, @3xp0n3nt and @relu91 have good points!

There is not much reason to not use the BatchRenderDevice for everything since it's a far better approach to rendering when considering any OpenGL backend.

However, I think @relu91 has a valid view as well. The original RenderDevice interface is so brutal simple that I'm a bit afraid that forcing everyone down the BatchRenderDevice way of doing things might be limiting in some corner cases. The BatchRenderDevice was only possible because of the simplicity of the original RenderDevice -- since it's being build on top of it. And I don't see a real reason to change that?

So, yes, it's a bit difficult to keep both but I'd vote for that anyway: Keep the original RenderDevice interface (just in case...) **BUT** kill all the existing implementations in favor of BatchRenderBackend implementations using the BatchRenderDevice! I totally agree with you @3xp0n3nt on that one! There is no reason to keep the old RenderDevice implementations!

And yes, going that way would include a Java2D implementation as well. I think this is what @bgroenks96 suggested as well, make a BatchRenderBackend-Implementation for Java2D -- which I really like! :smile:

Again, I really like the simplicity of the original RenderDevice which is the view that Nifty has when it needs to render stuff. And for Nifty 1.x I'd want to keep that, if possible.

**Comment by ghost** [comment]
@bgroenks96 What you are describing regarding simulating texture atlases sounds great. It is, however, confirming what I mentioned above about having to rewrite / duplicate all of the functionality in the batch rendering system, since you can't "tap into" the existing functionality, which is a shame. It's just going to be a lot more work for you, especially as the batch rendering system evolves, if you want to maintain feature parity.

**Comment by ghost** [comment]
@void256 Ok, so we'll leave `RenderDevice` alone, for now ;-).

(`RenderDevice` breathes a sigh of relief...). :sweat_smile:

**Comment by bgroenks96** [comment]
@3xp0n3nt More work, yes perhaps. But that doesn't mean I won't do it ;-P

If you can give me a general idea of what I should start working now, refactoring wise, I will get started as soon as possible. Should I start with the Batch rendering simulation?

**Comment by ghost** [comment]
> bgroenks96 commented 4 hours ago
> 
> Should I start with the Batch rendering simulation?

Sounds good to me; I don't see why not.

**Comment by bgroenks96** [comment]
@3xp0n3nt 

Exactly how does the batch renderer handle batched GUI draw requests? It writes them to a TextureAtlas correct? And this atlas is stored somewhere in VRAM? I suppose then the only actual performance benefit comes from not needing to rebind textures, since the data is still all being sent to the graphics card with each call. I'll have to think more about how to mimick that kind of capability in Java2D.

Do we want to put the batch renderer implementation in a new 'batch' sub-package or just scrap everything currently in the Java2D renderer and start over?

**Comment by ghost** [comment]
@bgroenks96 Your explanation is _basically_ correct, but you should read the javadoc of [BatchRenderDevice](https://github.com/void256/nifty-gui/blob/1.4/nifty-core/src/main/java/de/lessvoid/nifty/render/batch/BatchRenderDevice.java), [BatchRenderBackend](https://github.com/void256/nifty-gui/blob/1.4/nifty-core/src/main/java/de/lessvoid/nifty/render/batch/spi/BatchRenderBackend.java), & [BatchRenderConfiguration](https://github.com/void256/nifty-gui/blob/1.4/nifty-core/src/main/java/de/lessvoid/nifty/render/batch/BatchRenderConfiguration.java) to get more details. I was writing a wall of text when I realized that I've already done that in the above javadoc. The 'batch' packages are going away. I'd scrap everything and start over.

**Comment by bgroenks96** [comment]
We're keeping RenderDevice right? So wouldn't we want to keep the
RenderDevice implementation?

On Wed, Sep 24, 2014 at 3:41 PM, 3xp0n3nt notifications@github.com wrote:

> @bgroenks96 https://github.com/bgroenks96 Your explanation is
> _basically_ correct, but you should read the javadoc of BatchRenderDevice
> https://github.com/void256/nifty-gui/blob/1.4/nifty-core/src/main/java/de/lessvoid/nifty/render/batch/BatchRenderDevice.java,
> BatchRenderBackend
> https://github.com/void256/nifty-gui/blob/1.4/nifty-core/src/main/java/de/lessvoid/nifty/render/batch/spi/BatchRenderBackend.java,
> & BatchRenderConfiguration
> https://github.com/void256/nifty-gui/blob/1.4/nifty-core/src/main/java/de/lessvoid/nifty/render/batch/BatchRenderConfiguration.java
> to get more details. I was writing a wall of text when I realized that I've
> already done that in the above javadoc. The 'batch' packages are going
> away. I'd scrap everything and start over.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/void256/nifty-gui/issues/284#issuecomment-56726766.

## 

Brian Groenke
bgroe8@gmail.com

**Comment by ghost** [comment]
@bgroenks96 No, all non-batched `RenderDevice` implementations are deprecated and soon to be removed in an upcoming PR. It should stick around (for @relu91) at least until your new implementation is stable though.

**Comment by bgroenks96** [comment]
Does anyone (@relu91 @3xp0n3nt) have any suggestions for how bitmap font support could be implemented?

**Comment by relu91** [comment]
When I looked for font support some months ago, since the actual java2d render simply render one font using AWT fonts. What I have found is this two resources : 
- https://www.youtube.com/watch?v=uEOK8ihjDtU.
- http://www.java-gaming.org/index.php?action=pastebin&hex=54a6b5d74

Maybe you have just find them on your own, but still. :+1:

**Comment by bgroenks96** [comment]
Are you aware of GraphicsEnvironment.createFont? That lets you load off/ttf files. I've never worked directly with bitmap fonts so I'm not really sure what you're looking for.

**Comment by relu91** [comment]
Yes; I am. The problem is that some other renders support only bitmap fonts. If nifty editor ( and the java2d render) will only support tff fonts the gui generated with it will not be compatible with the other renders. That's why a searched to implement a renderer for bitmap font in java2d :smiley:

**Comment by bgroenks96** [comment]
Ok. I will look into it and see what I can do, but my first priority is getting the Java2D implementation of BatchRenderBackend up and running.

**Comment by bgroenks96** [comment]
Just as a quick update to everyone interested in this issue, a lot of the implementation code has been written, but the batch renderer is not currently in a working state. There are a lot of problems arising with the translation between the Java2D and OpenGL coordinate systems (especially the sort of awkward back converting from normalized texture coords).

Feel free to review the branch's current progress on [my fork repo](https://github.com/bgroenks96/nifty-gui/tree/java2d-batch-renderer).
